File: cpp/examples/linux/accelerometer/ADXL362/minimal/main.cpp
#include <cstdio>
#include <unistd.h>
#include "SPIConnectionLinux.h"
#include "ADXL362.h"

#ifndef TEST_SPI_BUS
#define TEST_SPI_BUS 0
#endif
#ifndef TEST_SPI_DEVICE
#define TEST_SPI_DEVICE 0
#endif

int main() {
    SPIConnectionLinux connection(TEST_SPI_BUS, TEST_SPI_DEVICE, 0, 8000000);    // Create SPI connection, (bus=0, device=0, mode=0, max_speed_hz=8e6) → SPIConnectionLinux
    ADXL362Minimal accel(connection);                                              // Create ADXL362 driver, (connection) → ADXL362Minimal

    for (;;) {
        float x, y, z;
        accel.read(x, y, z);                                                       // Read 3-axis acceleration, (x, y, z) → g, g, g
        printf("x=%+.3f  y=%+.3f  z=%+.3f g\n", x, y, z);
        usleep(100000);
    }
    return 0;
}

File: rust/examples/linux/accelerometer/adxl362/minimal/src/main.rs
use embedded_hal_bus::spi::ExclusiveDevice;
use linux_embedded_hal::SpidevBus;
use periph::chips::accelerometer::Adxl362Minimal;
use spidev::{SpiModeFlags, Spidev, SpidevOptions};

struct NullCs;
impl embedded_hal::digital::ErrorType for NullCs {
    type Error = core::convert::Infallible;
}
impl embedded_hal::digital::OutputPin for NullCs {
    fn set_low(&mut self) -> Result<(), Self::Error> { Ok(()) }
    fn set_high(&mut self) -> Result<(), Self::Error> { Ok(()) }
}

fn main() {
    let bus: u8 = std::env::var("SPI_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(0);
    let dev: u8 = std::env::var("SPI_DEVICE").ok().and_then(|v| v.parse().ok()).unwrap_or(0);
    let dev_path = format!("/dev/spidev{}.{}", bus, dev);

    let mut spi = Spidev::open(dev_path).expect("open spi");
    spi.configure(&SpidevOptions::new()
        .max_speed_hz(8_000_000)
        .mode(SpiModeFlags::SPI_MODE_0)
        .build()).expect("configure spi");
    let bus_obj = SpidevBus(spi);
    let device = ExclusiveDevice::new_no_delay(bus_obj, NullCs).expect("spi device");

    let mut accel = Adxl362Minimal::new(device).expect("init ADXL362");    // Create ADXL362 driver, (spi) → Result

    for _ in 0..10 {
        let (x, y, z) = accel.read().expect("read acceleration");            // Read 3-axis acceleration, () → (f32, f32, f32) g
        println!("x={:+.3}  y={:+.3}  z={:+.3} g", x, y, z);
        std::thread::sleep(std::time::Duration::from_millis(100));
    }
}